azure/services/securitygroups/securitygroups.go

```diff
@@ -83,7 +95,55 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	// We go through the list of security groups to reconcile each one, independently of the result of the previous one.
 	// If multiple errors occur, we return the most pressing one.
 	//  Order of precedence (highest -> lowest) is: error that is not an operationNotDoneError (i.e. error creating) -> operationNotDoneError (i.e. creating in progress) -> no error (i.e. created)
-	for _, nsgSpec := range specs {
+	for _, resourceSpec := range specs {
+		nsgSpec, ok := resourceSpec.(*NSGSpec)
+		if !ok {
+			return errors.New("cannot convert network security group spec")
+		}
+
+		newAnnotation := map[string]interface{}{}
+
+		// We only want to update annotations for the security rules that are always managed by CAPZ.
+		for _, securityRuleSpec := range nsgSpec.SecurityRulesSpecs {
+			if _, alwaysManaged := securityRuleLastAppliedAnnotation[securityRuleSpec.Annotation]; !alwaysManaged {
+				continue
+			}
+			newAnnotation[securityRuleSpec.SecurityRule.Name] = securityRuleSpec.SecurityRule
+		}
+
+		// Retrieve the last applied security rules for all NSGs.
+		lastAppliedSecurityRulesAll, err := s.Scope.AnnotationJSON(azure.SecurityRuleLastAppliedAnnotation)
+		if err != nil {
+			return err
+		}
+
+		// Retrieve the last applied security rules for this NSG.
+		lastAppliedSecurityRules, ok := lastAppliedSecurityRulesAll[nsgSpec.Name].(map[string]interface{})
+		if !ok {
+			lastAppliedSecurityRules = map[string]interface{}{}
+		}
+
+		// Delete the security rules were removed from the spec.
+		for ruleName := range lastAppliedSecurityRules {
+			if _, ok := newAnnotation[ruleName]; !ok {
+				_, err := s.client.DeleteRule(ctx, nsgSpec, ruleName)
+				if err != nil {
+					// if the security rule is already being deleted, we can ignore the error
+					if !azure.ResourceNotFound(err) {
+						return errors.Wrapf(err, "failed to delete security rule %s", ruleName)
```

Comment:
I think it would make sense to place the errors relating to operation not done at a higher priority than deleting security rules. Will update the logic accordingly.